AutoImplement/Program.cs

```diff
@@ -0,0 +1,193 @@
+
+            var genericInformation = type.Name.Contains("`") ? "`" + type.Name.Split('`')[1] : string.Empty;
+            var mainName = type.Name.Split('`')[0];
+            if (mainName.StartsWith("I")) mainName = mainName.Substring(1); // strip leading 'I' from interface name
+
+            GenerateImplementation<StubBuilder>     (type, $"Stub{mainName}{genericInformation}.cs",      "System.Delegation");
+            GenerateImplementation<CompositeBuilder>(type, $"Composite{mainName}{genericInformation}.cs", "System.Linq");
+            GenerateImplementation<DecoratorBuilder>(type, $"{mainName}Decorator{genericInformation}.cs");
+         }
+
+         Console.WriteLine($"Done generating implementations from {typeNames.Length} interfaces.");
+      }
+
+      private static void PrintUsageInformation() {
+         Console.WriteLine("Expected usage: AutoImplement <assembly> <type> <type> <type> ...");
+         Console.WriteLine("If no types are included, implementatiosn will be created for every public interface.");
+         Console.WriteLine("Example usage:");
+         Console.WriteLine("   AutoImplement \"mscorlib, Version=4.0.0.0, Culture=neutral, PublicKeyToken=b77a5c561934e089\" IDisposable");
+         Console.WriteLine("   AutoImplement path/to/my/assembly.dll IMyType");
+         Console.WriteLine("   <drag an assembly onto the program>");
+         Console.WriteLine("");
+         Console.WriteLine("Assembly name can either be fully qualified from the GAC or a file path.");
+         Console.WriteLine("Some common assemblies are preloaded so you can refer to them by their short name.");
+         Console.WriteLine("Namespaces on types are optional.");
+      }
+
+      /// <see cref="https://stackoverflow.com/questions/2531837/how-can-i-get-the-pid-of-the-parent-process-of-my-application/2533287#2533287"/>
+      private static void PauseIfStartedByExplorer() {
+         var myId = Process.GetCurrentProcess().Id;
+         var query = $"SELECT ParentProcessId FROM Win32_Process WHERE ProcessId = {myId}";
+         var search = new ManagementObjectSearcher("root\\CIMV2", query);
+         var results = search.Get().GetEnumerator();
+         results.MoveNext();
+         var queryObj = results.Current;
+         var parentId = (uint)queryObj["ParentProcessId"];
+         var parent = Process.GetProcessById((int)parentId);
+
+         if (parent.ProcessName == "explorer") {
+            Console.ReadKey();
+         }
+      }
+
+      /// <summary>
+      /// The user could've passed us a string representing the long name of an assembly in the GAC.
+      /// They also could've passed us a string representing a file path on disk.
+      /// We don't know which it is, so just try both.
+      /// </summary>
+      private static bool TryGetAssembly(string candidate, out Assembly result) {
+         if (TryGetAssemblyFromLongName(candidate, out result) || TryGetAssemblyFromFile(candidate, out result)) {

```

Comment:
I know the TryGet pattern allows you to short circuit like this, but the methods are now doing to different things. On top of that, debugging this method (without stepping into lower methods) is going to be a pane because you won't know which of the TryGetAssembly... methods yielded the result.

Why not (changing the sub methods to return null on failure)
```
result = TryGetAssemblyFromLongName(candidate);
if (result != null) {
  return result;
}

result = TryGetAssemblyFromFile(candidate) ;
if(result != null){
  return result;
}

// else
PrintUsageInformation(); // actually - this should probably not be in here either . . .
return null;
```

Reply:
I'm not willing to break the .net convention of try methods returning bools, but I did reorganize the code a bit based on your suggestion to make it easier to tell what's going on.